### PR TITLE
Handle a special use-case of user home directory permissions

### PR DESCRIPTION
**Pull request by favoretti**
## Changes
In a special case when one would want to add rights for extra users to either view or edit things in one's home directory - the user's own `CAN_MANAGE` permission is implicit. 

One can specify it explicitly in the resource, however, during deletion that `CAN_MANAGE` permission can not be removed, resulting in a failed run that can't be fixed without deleting resource from the state (which leaves other permissions hanging).

Hence here's an attempt to ignore or implicitly add it.

This works when permission is supplied with `directory_path` that matches the `user_name`, however, this also doesn't work when we specify `user_id`... 

Therefore as WIP for now..

## Tests

- [x] `make test` run locally
- [ ] relevant change in `docs/` folder
- [ ] covered with integration tests in `internal/acceptance`
- [ ] relevant acceptance tests are passing
- [ ] using Go SDK

### CODE REVIEW

File: permissions/resource_permissions.go [hunk 1]
```diff
@@ -215,6 +215,16 @@ func (a PermissionsAPI) Delete(objectID string) error {
 				accl.AccessControlList = append(accl.AccessControlList, change)
 			}
 		}
+
+		// handle special case when we add extra permission to a user home dir
+		if v, ok := d.GetOk("directory_path"); ok {
```
Comment (alexott) [inline permissions/resource_permissions.go:220]: We need to think how we will handle this for the case when we specify object ID instead of directory path...

**Review by alexott — COMMENTED** [review]
We also need to test how it will behave if we'll try to change permissions for user/SP from manage to lower, like, CAN_READ...

File: permissions/resource_permissions.go (hunk 1 above)
Comment (favoretti) [reply permissions/resource_permissions.go:220]: Yeah, that's a tricky one my brain refused to wrap itself around at 3am.. Directory ID won't match `user_name` of course, but fetching a user record based on a user id seemed like an overkill. Open for any hints here.

**Comment by favoretti** [comment]
> We also need to test how it will behave if we'll try to change permissions for user/SP from manage to lower, like, CAN_READ...

This particular use-case is only intended for a single corner-case permission: user `X` to `/Users/X`, so just this particular user's home dir. That case of `CAN_MANAGE` is implicit, you can't lower it I believe.

**Comment by alexott** [comment]
I'm clarifying with the team who is responsible for workspace-level permissions

File: permissions/resource_permissions.go (hunk 1 above)
Comment (alexott) [reply permissions/resource_permissions.go:220]: yes, it's possible to get the directory path by object iD

File: permissions/resource_permissions_test.go
Comment (alexott) [inline permissions/resource_permissions_test.go:1]: I think that we'll need to expand tests to cover this stuff.

File: permissions/resource_permissions.go (hunk 1 above)
Comment (alexott) [inline permissions/resource_permissions.go:220]: We may need to disallow changes only for the current user - then we'll need to add a check like `if me == accessControl.UserName || me == accessControl.ServicePrincipalName {` like [below](https://github.com/databricks/terraform-provider-databricks/pull/3586/files#diff-e21ed6630273de6b0cb95643ddfc8cda8082f1ed5c5ce39b53210622ad90f55aR339). We also need to handle the case of home directory of the service principal, not only ordinary users.

**Comment by alexott** [comment]
I got confirmation from the dev team:

> Currently, the `DirectoryPermissionsHandler` prevents users from changing the `CAN_MANAGE` permissions of a user on its own home folder.

**Comment by favoretti** [comment]
@alexott So, to summarize what I need to do to get this merged:
1. Add SPN home directories to the use-case
2. Handle `object_id` next to currently implemented `directory_path`
3. Write tests

Correct?

**Comment by alexott** [comment]
yes, it looks like. @mgyucht I think it makes sense to discuss in the next office hours

File: permissions/resource_permissions.go [hunk 2]
```diff
@@ -215,6 +215,16 @@ func (a PermissionsAPI) Delete(objectID string) error {
...
+			if v.(string) == fmt.Sprintf("/Users/%s", acl.UserName) && acl.AllPermissions[0].PermissionLevel == "CAN_MANAGE" {
```
Comment (tanmay-db) [inline permissions/resource_permissions.go:221]: Why do we only compare with the first index of the list? Just to confirm, is it necessar that CAN_MANAGE will always be at 0 index?

**Review by tanmay-db — COMMENTED** [review]
The logic looks good to me overall, I had a concern on checking CAN_MANAGE with first index of the list.

File: permissions/resource_permissions.go (hunk 2 above)
Comment (favoretti) [reply permissions/resource_permissions.go:221]: @tanmay-db That is... a great question, considering that this PR has been hanging here for a while - I'm not sure... I'm unfortunately not that knee-deep in databricks provider code, would you mind suggesting an alternative?

Comment (tanmay-db) [reply permissions/resource_permissions.go:221]: @alexott and me discussed this and we think going ahead with checking all the items in the list would be more safer.

Comment (tanmay-db) [reply permissions/resource_permissions.go:221]: Just to check, after this extending this to SPN and adding tests would be left before we can merge this right?

Comment (favoretti) [reply permissions/resource_permissions.go]: @tanmay-db Something like this?

Comment (favoretti) [reply permissions/resource_permissions.go]: > Just to check, after this extending this to SPN and adding tests would be left before we can merge this right?

Expanding to SPN homedir you mean? Yeah, I'll need to work on that too, although I have no ideas what the format of SPN's home dir is... is it `/Users/<spn_id>`? If so, which ID is it, Azure Object ID, Azure app ID or Databricks internal ID? :)

As for tests... I'll do my best, but would really appreciate some help, I'm incredibly new to your codebase :)

Comment (alexott) [reply permissions/resource_permissions.go]: it's SPN's Application ID (UUID)

Comment (tanmay-db) [reply permissions/resource_permissions.go]: @favoretti, yes, to optimise a little, we can add break in the loop when we find the permission level to CAN_MANAGE.
